gemrb/core/Cache.h

```diff
@@ -29,65 +31,94 @@ namespace GemRB {
 using ReleaseFun = void (*)(void*);
 #endif
 
-class Cache
-{
-protected:
-	// Association
-	struct MyAssoc {
-		MyAssoc* pNext;
-		MyAssoc** pPrev;
-		ResRef key;
-		ieDword nRefCount;
-		void* data;
-	};
-	struct MemBlock {
-		MemBlock* pNext;
-	};
-
-public:
-	// Construction
-	explicit Cache(int nBlockSize = 10, int nHashTableSize = 129);
-	Cache(const Cache&) = delete;
-	~Cache();
-	Cache& operator=(const Cache&) = delete;
-
-	// Attributes
-	// number of elements
-	inline int GetCount() const
-	{
-		return m_nCount;
-	}
-	inline bool IsEmpty() const
-	{
-		return m_nCount==0;
+/* Reference counting cache, a layer between STL containers and existing interfaces. */
+template<typename K, typename V, typename H>
+class RCCache {
+	private:
+		struct Value {
+			V value;
+			int64_t refCount;
+
+			Value (V && value) : value(std::forward<V>(value)), refCount(1) {}
+
+			template<typename ... ARGS>
+			Value (ARGS && ... args) : value(std::forward<ARGS>(args)...), refCount(1) {}
+		};
+
+		std::unordered_map<K, Value, H> map;
+
+	public:
+		V* GetResource(const K& key) {
+			auto lookup = map.find(key);
+			if (lookup != map.cend()) {
+				lookup->second.refCount++;
+
+				return &lookup->second.value;
+			}
+
+			return nullptr;
+		}
+
+		template<typename ... ARGS>
+		std::pair<V*, bool> SetAt(const K& key, ARGS && ... args) {
+			auto insertion =
+				map.emplace(
+					std::piecewise_construct,
+					std::forward_as_tuple(key),
+					std::forward_as_tuple(std::forward<ARGS>(args)...)
```

Comment:
These are defined in <tuple>, so it might be prudent to include it directly.

Reply:
Yes, won't do any harm.